File: 560-subarray-sum-equals-k/560-subarray-sum-equals-k.cpp
class Solution {
public:

    int subarraySum(vector<int>& nums, int k) {
    
        int res=0;
        map<int, int> m;
        m[0]++;
        int cursum=0;
        for(int i=0; i<nums.size(); i++)
        {
            cursum+=nums[i];
            int diff = cursum - k;
            if(m.find(diff) != m.end())
                res+=m[diff];
            
            m[cursum]++;
        }
        return res;
    }    
};

File: 560-subarray-sum-equals-k/560-subarray-sum-equals-k.java
class Solution {
  public int subarraySum(int[] nums, int k) {
    if (nums == null || nums.length == 0) {
        return 0;
    }
    
    return subarraySum(nums, k, 0);
}

public int subarraySum(int[] nums, int k, int from) {
    if (nums.length <= from) {
        return 0;
    }
    
    int result = 0;
    
    int sum = 0;
    for (int i = from; i < nums.length; i++) {
        sum += nums[i];
        
        if (sum == k) {
            result += 1;
        }
    }
    
    return result + subarraySum(nums, k, from + 1);
}

}